sql/core/src/main/scala/org/apache/spark/sql/artifact/ArtifactManager.scala

```diff
@@ -279,6 +296,49 @@ class ArtifactManager(session: SparkSession) extends Logging {
     loader
   }
 
+  private[sql] def clone(newSession: SparkSession): ArtifactManager = {
+    val sparkContext = session.sparkContext
+    sparkContext.synchronized {
+      val newArtifactManager = new ArtifactManager(newSession)
+      if (artifactPath.toFile.exists()) {
+        FileUtils.copyDirectory(artifactPath.toFile, newArtifactManager.artifactPath.toFile)
+      }
+      val blockManager = sparkContext.env.blockManager
+      val newBlockIds = cachedBlockIdList.asScala.map { blockId =>
+        val newBlockId = blockId.copy(sessionUUID = newSession.sessionUUID)
+        copyBlock(blockId, newBlockId, blockManager)
+      }
+
+      // Re-register resources to SparkContext
```

Comment:
Does this copy or just register? If it does the latter, we could also consider creating an explicit method in SparkContext for this.

Reply:
Can you file a follow-up that tries to make this better?

Reply:
Will do.